python/ray/autoscaler/cli_logger.py

```diff
@@ -0,0 +1,426 @@
+        if rendered_message and rendered_message[-1] != "\n":
+            rendered_message += " "
+
+        l = len(rendered_message.split("\n")[-1])
+        complete_str = rendered_message + confirm_str
+
+        if yes:
+            self._print(
+                complete_str + "y " + cf.gray("[automatic, due to --yes]"))
+            return True
+
+        self._print(complete_str, linefeed=False)
+
+        res = None
+        yes_answers = ["y", "yes", "true", "1"]
+        no_answers = ["n", "no", "false", "0"]
+        try:
+            while True:
+                ans = sys.stdin.readline()
+                ans = ans.lower()
+
+                if ans == "\n":
+                    res = default
+                    break
+
+                ans = ans.strip()
+                if ans in yes_answers:
+                    res = True
+                    break
+                if ans in no_answers:
+                    res = False
+                    break
+
+                indent = " " * l
+                self.error(
+                    "{}Invalid answer: {}. "
+                    "Expected {} or {}",
+                    indent, cf.bold(ans.strip()),
+                    self.render_list(yes_answers, "/"),
+                    self.render_list(no_answers, "/"))
+                self._print(indent + confirm_str, linefeed=False)
+        except KeyboardInterrupt:
+            self.newline()
+            res = default
+
+        if not res and should_abort:
+            # todo: make sure we tell the user if they
+            # need to do cleanup
+            self._print("Exiting...")
+            sys.exit(2)
+
+        return res
+
+    def old_confirm(self, msg, yes):
+        if not self.old_style:
+            return
+
+        return None if yes else click.confirm(msg, abort=True)
+
+    def boto_exception_handler(self, msg, *args, **kwargs):
+        # todo: implement timer
+        cli_logger = self
+        class ExceptionHandlerContextManager():
+            def __enter__(self):
+                pass
+
+            def __exit__(self, type, value, tb):
+                import botocore
+
+                if type is botocore.exceptions.ClientError:
+                    cli_logger.handle_boto_error(value, msg, *args, **kwargs)
+
+        return ExceptionHandlerContextManager()
+
+    def handle_boto_error(self, exc, msg, *args, **kwargs):
+        if self.old_style:
+            # old-style logging doesn't do anything here
+            return
+
+        error_code = None
+        error_info = None
+        # todo: not sure if these exceptions always have response
+        if hasattr(exc, "response"):
+            error_info = exc.response.get("Error", None)
+        if error_info is not None:
+            error_code = error_info.get("Code", None)
+
+        generic_message_args = [
+            "{}\n"
+            "Error code: {}",
+            msg.format(*args, **kwargs),
+            cf.bold(error_code)
```

Comment:
we will not want to have boto-specific logic here right? this should just be a helper function in autoscaler/aws

Reply:
yeah should move this out to a utility file